app/models/robot_directory.rb

```diff
@@ -1,59 +1,101 @@
+      'robohash', robot[:birthdate], robot[:date_hired], robot[:department])
+
+    serial = directory.execute("SELECT serial_number
+                                      FROM robots
+                                      WHERE name=?;", robot[:name]).flatten.last['serial_number']
+    directory.execute("UPDATE robots
+                      SET avatar='https://robohash.org/#{serial}'
+                      WHERE name=?;", robot[:name])
   end
 
   def all
-    directory.transaction do
-      robots.map { |robot| Robot.new(robot) }
-    end
+    raw_robots.map { |robot| Robot.new(robot) }
   end
 
-  def robots
-    directory.transaction do
-      database['robots'] || []
-    end
+  def raw_robots
+    directory.execute("SELECT * FROM robots;")
   end
 
-  def robot(serial_number)
-    robots.find { |robot| robot['serial_number'] == serial_number }
+  def raw_robot(serial_number)
+    directory.execute("SELECT * FROM robots
+                       WHERE serial_number=?;", serial_number).first
+  end
+
+  def find(serial_number)
+    Robot.new(raw_robot(serial_number))
   end
 
   def update(serial_number, robot)
-    directory.transaction do
-      target_robot = directory['robots'].find { |robot| robot['serial_number'] == serial_number }
-      target_robot['name']       = robot[:name]
-      target_robot['city']       = robot[:city]
-      target_robot['state']      = robot[:state]
-      target_robot['avatar']     = robot[:avatar]
-      target_robot['birthdate']  = robot[:birthdate]
-      target_robot['date_hired'] = robot[:date_hired]
-      target_robot['department'] = robot[:department]
-    end
+    directory.execute("UPDATE robots
+                       SET name=?, city=?, state=?, birthdate=?, date_hired=?,
+                       department=? WHERE serial_number=?;", robot[:name],
+                       robot[:city], robot[:state], robot[:birthdate],
+                       robot[:date_hired], robot[:department], serial_number)
   end
 
   def delete(serial_number)
-    directory.transaction do
-      directory['robots'].delete_if { |robot| robot['serial_number'] == serial_number }
-    end
+    directory.execute("DELETE FROM robots WHERE serial_number=?;", serial_number)
+  end
+
+  def average_age
+    ages = all.map { |robot| Time.now.year - Time.new(robot.birthdate).year }
+    ages.reduce(:+) / ages.count
+  end
+
+  def hired_by_year
+    years = all.map { |robot| Time.new(robot.date_hired).year }.sort
+    unique_years = Hash.new(0)
+    years.each { |year| unique_years[year] += 1 }
+    unique_years
+  end
+
+  def by_department
+    departments = all.map { |robot| robot.department }.sort
+    unique_departments = Hash.new(0)
+    departments.each { |department| unique_departments[department] += 1 }
+    unique_departments
+  end
+
+  def by_city
+    cities = all.map { |robot| robot.city }.sort
+    unique_cities = Hash.new(0)
+    cities.each { |city| unique_cities[city] += 1 }
+    unique_cities
+  end
+
+  def by_state
```

Comment:
this method looks pretty similar to `by_department` and `by_city`. is there a way you can think of to avoid writing so much code and maybe encapsulate this behavior in a single method that takes an argument?